src/foam/parse/DateParser.js

```diff
@@ -0,0 +1,727 @@
+        if ( expectedYear !== undefined && expectedMonth !== undefined && expectedDay !== undefined ) {
+          let actualYear = date.getFullYear();
+          let actualMonth = date.getMonth();
+          let actualDay = date.getDate();
+
+          if ( actualYear !== expectedYear || actualMonth !== expectedMonth || actualDay !== expectedDay ) {
+            // Date was normalized (e.g., 2025-13-01 became 2026-01-01)
+            date = foam.Date.MAX_DATE;
+            console.warn("Invalid date: " + str + "; assuming " + date.toISOString() + ".");
+          }
+        }
+
+        return date;
+      }
+    },
+
+    {
+      name: 'validateDateUTC',
+      documentation: 'Validates a date object (UTC time) and returns MAX_DATE for invalid dates',
+      code: function(date, str, expectedYear, expectedMonth, expectedDay) {
+        // Check if date is NaN
+        if ( isNaN(date.getTime()) ) {
+          date = foam.Date.MAX_DATE;
+          console.warn("Invalid date: " + str + "; assuming " + date.toISOString() + ".");
+          return date;
+        }
+
+        // If we have expected values, verify the date wasn't normalized by JS Date constructor
+        if ( expectedYear !== undefined && expectedMonth !== undefined && expectedDay !== undefined ) {
+          let actualYear = date.getUTCFullYear();
+          let actualMonth = date.getUTCMonth();
+          let actualDay = date.getUTCDate();
+
+          if ( actualYear !== expectedYear || actualMonth !== expectedMonth || actualDay !== expectedDay ) {
+            // Date was normalized (e.g., 2025-13-01 became 2026-01-01)
+            date = foam.Date.MAX_DATE;
+            console.warn("Invalid date: " + str + "; assuming " + date.toISOString() + ".");
+          }
+        }
+
+        return date;
+      }
+    },
+
+    {
+      name: 'convertTwoDigitYear',
+      documentation: 'Converts 2-digit year using fixed pivot: 00-49 → 2000-2049, 50-99 → 1950-1999',
+      code: function(twoDigitYear) {
+        // Fixed pivot at 50:
+        // Years 00-49 map to 2000-2049
+        // Years 50-99 map to 1950-1999
+        if ( twoDigitYear < 50 ) {
+          return 2000 + twoDigitYear;
+        } else {
+          return 1900 + twoDigitYear;
+        }
+      }
+    },
+
+    {
+      name: 'parseString',
+      documentation: 'Parse a date/datetime string and return a Date object. Auto-detects format and handles time if present. Returns MAX_DATE for invalid dates.',
+      code: function(str, opt_name) {
+        let result = this.grammar_.parseString(str, opt_name || 'START');
+
+        if ( ! result ) {
+          // Unparseable format - return MAX_DATE
+          return this.validateDate(new Date(NaN), str);
+        }
+
+        // Determine if this is a datetime or date-only result based on presence of time components
+        let ret;
+        if ( result.hour !== undefined || result.minute !== undefined || result.second !== undefined ) {
```

Comment:
Each of the parser actions knows what kind of date they're working with, so wouldn't it have been better if they had just created the date in the correct format in the first place rather than having to create twice as many Dates?

Reply:
no, since we have parseDate, parseDateTime, and parseDateTimeUtc, so I send the result as an object and based on the function I decide noon(hours), vs UTC, for example